### ISSUE TITLE
How to download the grid as PDF file ?

### ISSUE THREAD

**Issue by osamakamel** [optikalefx/OpenJS-Grid#82]
Hello Sean
First of all thanks for this great work ( i appreciate this work too much)
But can you tell me is there a way to download the grid as PDF file?
Any help would be greatly appreciated
Thanks

**Comment by optikalefx** [comment]
Hi,  I don't have any functionality for that as I've never needed it. You're welcome to contribute to the project if you'd like.

**Comment by osamakamel** [comment]
Thanks Sean for your very fast answer.I'll be working on this as i need this in my work so much (i would be pleased to contribute to your work if i can)

**Comment by optikalefx** [comment]
Awesome. The way I would go about solving this would be to pass `FALSE` in for the last parameter of the grid call. This then gives you access to the entire set of data rows. You should then easily be able to pass that into any PDF parser. HTML2PDF or WK2PDF.

```php
$grid = new Grid("user", array(
	"joins"=>array("LEFT JOIN activation ON (user.id = activation.user_id)"),
	"fields"=>array(
		"name"=>"CONCAT(first_name,' ',last_name)",
		"first_name"=>"user.first_name",
		"last_name"=>"user.last_name",
		"activated"=>"activation.used",
		"pass"=>"''"
	),
	"having"=>"first_name IS NOT NULL && 1=1"
),FALSE);

// loop over data
foreach($grid->data['rows'] as $key=>$row) {

}

**Comment by osamakamel** [comment]
Sorry Sean but i don't get it 
but when i use $row in HTML2PDF parser i got the PDF file with the string "array" instead of the data  and it is only for the column shown on the page not the whole grid

**Comment by optikalefx** [comment]
So `$row` is an associative array. For instance in the example above you would get the name by saying `$row['name']`.

The data will be for exactly what you select in the `new Grid` call.  If you want it to match your grid exactly, then make the exact same `new Grid` call.

**Comment by osamakamel** [comment]
Thanks Sean . Now it works fine and i have the data in the PDF file but is there a way to get all the data in the entire grid not only the data displayed in the page?
I appreciate your help.

**Comment by optikalefx** [comment]
when you say "entre grid" do you mean entire database?

**Comment by osamakamel** [comment]
I mean the entire data (result) that is matched with my SQL query

**Comment by optikalefx** [comment]
So this one is a little weird. Basically you need to set `$_POST['nRowsShowing'] = 500;` or whatever number is greater than the number of rows you need.  This will fool the system into thinking you want more rows requested.

**Comment by osamakamel** [comment]
I tried this solution (changing the nRowsShowing in the html/php file ) before. but i needed a better solution to call the entire data without changing the displayed page. is there a way or it is inordinate? 
(excuse  me for my bad language)

**Comment by optikalefx** [comment]
Well technically the call to make  PDF would be a different call than the one to create the grid. So you shouldn't be running into that problem.  If you are running into problems, you could try to `unset($_POST['nRowsShowing']);` before you're done with the PDF

**Comment by osamakamel** [comment]
I am sorry to be too Greedy but if only I could get a comprehensive example where exactly to use this :  unset($_POST['nRowsShowing']);
$_POST['nRowsShowing'] = 500;
 , I wouldn't be bothering you.

**Comment by optikalefx** [comment]
```php

$_POST['nRowsShowing'] = 500;

$grid = new Grid("user", array(
    "joins"=>array("LEFT JOIN activation ON (user.id = activation.user_id)"),
    "fields"=>array(
        "name"=>"CONCAT(first_name,' ',last_name)",
        "first_name"=>"user.first_name",
        "last_name"=>"user.last_name",
        "activated"=>"activation.used",
        "pass"=>"''"
    ),
    "having"=>"first_name IS NOT NULL && 1=1"
),FALSE);

// loop over data
foreach($grid->data['rows'] as $key=>$row) {

}
```

**Comment by osamakamel** [comment]
I tried this before but it changed the rows displayed at the page (which i don't want)

**Comment by optikalefx** [comment]
I'm not sure how that's possible. If these are totally separate ajax calls,
one would not affect the other.  Can you give me more information how these
2 ajax calls might be interfering?

**Comment by osamakamel** [comment]
I would tell you what i was thinking about (the earlier question i asked was to know if the SQL query is executed one time and fetch all the data from the database then display it page by page according to the nRowsShowing OR it is executed every time i click next and fetching its content only) 
If you may cope with me , i thought if it is fetching all the data (SQL query runs one time only) then it is possible to manipulate the nRowsShowing at the PDF without changing them at the page displayed However it won't be possible at the other case ,SO i thought it would be possible to add the parser as a function at  grid.php file by passing the parameters  of the gathered SQL that is needed in the parsing 
(that was my way) you may fix it if it was wrong

**Comment by optikalefx** [comment]
You need a totally different ajax call to make this work. The paging
function makes a new ajax call each time. I would duplicate your ajax.php
and have a button call that new file which would output a PDF to the browser

**Comment by osamakamel** [comment]
How can i have a button call to the new file without making a double call for the grid?

**Comment by optikalefx** [comment]
You WANT to make a double call to the grid. This PDF script should be
totally separate from any code you have for the grid itself.

**Comment by osamakamel** [comment]
THANKS Sean for your help. I APPRECIATE THAT.

**Comment by optikalefx** [comment]
YOUR WELCOME!

**Comment by MGA555** [comment]
Is what osamakamel has done suitable to be added to OpenJS-Grid as a commit?  If not, how about creating a feature branch so that the community can take advantage of the work done?  

This would inspire others to contribute as well.  I would like to build on this to create an output to excel functionality.

**Comment by optikalefx** [comment]
The code to make this work is already part of the API. If @osamakamel want's to share his PDF code he can, but it's going to be very different for every user i'd imagine. Every PDF style will probably differ. The important bit is above where we show how to grab the grid data and do something with it.

**Comment by osamakamel** [comment]
I believe that Sean is right this is the most important part :
 " The important bit is above where we show how to grab the grid data and do something with it " .
The other work is all about the library I've used which is HTML2PDF and style .
But if you don't mind Sean , i think it would be possible to add this library to the grid files then add the call PDF function  to grid.php file as we could pass $cols (at line 196) variable instead of the foreach loop in the ajax file then we can pass the style from the ajax file to grid.php then it will be fixed in the grid. that would be my thoughts and it won't change the grid call .

**Comment by optikalefx** [comment]
The 2 biggest thing that the Grid needs at this point, which I still haven't found time to do, is upgrading to bootstrap 3 and using mysql PDO. Until those 2 are done, I don't think we should add to the grid file.  But we could still take a look at a PR to see what you mean above.
